pkg/scheduler/framework/plugins/interpodaffinity/interpod_affinity.go

```diff
@@ -55,39 +59,303 @@ var _ framework.FilterPlugin = &InterPodAffinity{}
+	topologyToMatchedAffinityTerms topologyToMatchedTermCount
+	// A map of topology pairs to the number of existing pods that match the anti-affinity terms of the "pod".
+	topologyToMatchedAntiAffinityTerms topologyToMatchedTermCount
 }
 
 // Clone the prefilter state.
 func (s *preFilterState) Clone() framework.StateData {
-	copy := &preFilterState{
-		meta: s.meta.Clone(),
+	if s == nil {
+		return nil
+	}
+
+	copy := preFilterState{}
+	copy.topologyToMatchedAffinityTerms = s.topologyToMatchedAffinityTerms.clone()
+	copy.topologyToMatchedAntiAffinityTerms = s.topologyToMatchedAntiAffinityTerms.clone()
+	copy.topologyToMatchedExistingAntiAffinityTerms = s.topologyToMatchedExistingAntiAffinityTerms.clone()
+
+	return &copy
+}
+
+// updateWithPod updates the preFilterState counters with the (anti)affinity matches for the given pod.
+func (s *preFilterState) updateWithPod(updatedPod, pod *v1.Pod, node *v1.Node, multiplier int64) error {
+	if s == nil {
+		return nil
 	}
+
+	// Update matching existing anti-affinity terms.
+	updatedPodAffinity := updatedPod.Spec.Affinity
+	if updatedPodAffinity != nil && updatedPodAffinity.PodAntiAffinity != nil {
+		antiAffinityProperties, err := getAffinityTermProperties(pod, schedutil.GetPodAntiAffinityTerms(updatedPodAffinity.PodAntiAffinity))
+		if err != nil {
+			klog.Errorf("error in getting anti-affinity properties of Pod %v", updatedPod.Name)
+			return err
+		}
+		s.topologyToMatchedExistingAntiAffinityTerms.updateWithAntiAffinityTerms(pod, node, antiAffinityProperties, multiplier)
+	}
+
+	// Update matching incoming pod (anti)affinity terms.
+	affinity := pod.Spec.Affinity
+	podNodeName := updatedPod.Spec.NodeName
+	if affinity != nil && len(podNodeName) > 0 {
+		if affinity.PodAffinity == nil {
+			affinityProperties, err := getAffinityTermProperties(pod, schedutil.GetPodAffinityTerms(affinity.PodAffinity))
+			if err != nil {
+				klog.Errorf("error in getting affinity properties of Pod %v", pod.Name)
+				return err
+			}
+			s.topologyToMatchedAffinityTerms.updateWithAffinityTerms(updatedPod, node, affinityProperties, multiplier)
+		}
+		if affinity.PodAntiAffinity != nil {
+			antiAffinityProperties, err := getAffinityTermProperties(pod, schedutil.GetPodAntiAffinityTerms(affinity.PodAntiAffinity))
+			if err != nil {
+				klog.Errorf("error in getting anti-affinity properties of Pod %v", pod.Name)
+				return err
+			}
+			s.topologyToMatchedAntiAffinityTerms.updateWithAntiAffinityTerms(updatedPod, node, antiAffinityProperties, multiplier)
+		}
+	}
+	return nil
+}
+
+// TODO(Huang-Wei): It might be possible to use "make(map[topologyPair]*int64)" so that
+// we can do atomic additions instead of using a global mutext, however we need to consider
+// how to init each topologyToMatchedTermCount.
+type topologyPair struct {
+	key   string
+	value string
+}
+type topologyToMatchedTermCount map[topologyPair]int64
```

Comment:
I don't like how we use `map[string]map[string]int64` for scoring, but `map[topologyPair]int64` for filtering. We should probably just pick one (the one that gives us the best performance). But that's for a follow up.

Reply:
yeah, priority and predicate evolved separately, make sense to unify them, let me see if I can do it in this PR.

Reply:
lets do that as a follow up